### PR TITLE
`Programming exercises`: Include practice participations in the participation table

### PR DESCRIPTION
**Pull request by JohannesStoehr**
#### General

- [x] I tested **all** changes and their related features with **all** corresponding user types on a test server.
- [x] This is a small issue that I tested locally and was confirmed by another developer on a test server.
- [x] Language: I followed the [guidelines for inclusive, diversity-sensitive, and appreciative language](https://docs.artemis.ase.in.tum.de/dev/guidelines/language-guidelines/).
- [x] I chose a title conforming to the [naming conventions for pull requests](https://artemis-platform.readthedocs.io/en/latest/dev/guidelines/development-process.html#naming-conventions-for-github-pull-requests).
#### Server
- [x] I followed the [coding and design guidelines](https://docs.artemis.ase.in.tum.de/dev/guidelines/server/).
- [x] I added multiple integration tests (Spring) related to the features (with a high test coverage).
- [x] I documented the Java code using JavaDoc style.
#### Client
- [x] I followed the [coding and design guidelines](https://docs.artemis.ase.in.tum.de/dev/guidelines/client/) and ensured that the layout is responsive.
- [x] I documented the TypeScript code using JSDoc style.
- [x] I translated all newly inserted strings into English and German.

### Motivation and Context

Instructors should be able to see the practice participations after the deadline without affecting the scores or the assessment of the exercise.

### Description

The practice participation are now included in the participation overview of the exercise. 
Additionally there is a filter to only show the graded participations. The filter is only visible after the deadline, since there cannot be any practice participations before the deadline.
The practice participations should not be visible in the submissions, the scores or the assessment view of the programming exercise.

### Steps for Testing

Prerequisites:
- 1 Instructor
- 2 Students
- 1 Programming Exercise with a deadline

1. Log in to Artemis.
2. Participate with each student before the deadline.
3. Participate after the deadline using the practice mode after the deadline.
4. Check during each step if the participation overview looks correct to the instructor.
5. Check that submissions, scores and assessments only show the graded participation.

#### Code Review
- [x] Review 1
- [x] Review 2
#### Manual Tests
- [x] Test 1
- [x] Test 2

### Test Coverage

| Class/File | Branch | Line |
|------------|-------:|-----:|
| ParticipationService.java | 78% | 91% |
| ParticipationResource.java | 66% | 90% |

### CODE REVIEW

File: src/test/java/de/tum/in/www1/artemis/service/SubmissionServiceTest.java [hunk 1]
```diff
@@ -158,16 +158,16 @@ private void getTutorSpecificCallsTutor1(Exercise exercise) {
         database.changeUser("tutor1");
         unassessedSubmissionCorrectionRound0Tutor1 = submissionService.getRandomAssessableSubmission(exercise, true, 0);
         unassessedSubmissionCorrectionRound1Tutor1 = submissionService.getRandomAssessableSubmission(exercise, true, 1);
-        submissionListTutor1CorrectionRound0 = submissionService.getAllSubmissionsAssessedByTutorForCorrectionRoundAndExercise(exercise.getId(), tutor1, true, 0);
-        submissionListTutor1CorrectionRound1 = submissionService.getAllSubmissionsAssessedByTutorForCorrectionRoundAndExercise(exercise.getId(), tutor1, true, 1);
+        submissionListTutor1CorrectionRound0 = submissionService.getAllSubmissionsAssessedByTutorForCorrectionRoundAndExerciseIgnoreTestRuns(exercise.getId(), tutor1, true, 0);
```
Comment (simon-hng) [inline src/test/java/de/tum/in/www1/artemis/service/SubmissionServiceTest.java:161]: I would prefer a shorter name like 
`getAllSubmissionsAssessedByTutorIgnoreTestRuns` as exerciseId and correction round are communicated through the parameters of the function

**Review by simon-hng — DISMISSED** [review]
Tested in testing session and code looks good

File: src/test/java/de/tum/in/www1/artemis/service/SubmissionServiceTest.java (hunk 1 above)
Comment (JohannesStoehr) [reply src/test/java/de/tum/in/www1/artemis/service/SubmissionServiceTest.java:161]: We currently use "ForCorrectionRound" in multiple methods, so I'd like to keep it consistent.
We might want to think about a refactoring PR that takes care of all these occurrences

**Review by julian-christl — COMMENTED** [review]
Tested on TS1.

On the general Assessment Dashboard, some submissions are counted wrongfully.
It's correct on the exercise Assessment Dashboard.
Should be 3/3:

**Review by ge65cer — COMMENTED** [review]
Suggestion: Maybe you can use icons for the false/true column like on other pages?
Screenshot 2022-10-13 at 13 32 34

**Review by stefanwaldhauser — CHANGES_REQUESTED** [review]
Tested it and everything worked except the issue mentioned by @julian-christl

**Review by stefanwaldhauser — DISMISSED** [review]
Reapprove now that the issues are fixed

File: src/main/java/de/tum/in/www1/artemis/service/ParticipationService.java [hunk 2]
```diff
@@ -350,7 +350,14 @@ public ProgrammingExerciseStudentParticipation resumeProgrammingExercise(Program
         participation = configureBuildPlan(participation);
         // Note: the repository webhook (step 1c) already exists, so we don't need to set it up again, the empty commit hook (step 2c) is also not necessary here
         // and must be handled by the calling method in case it would be necessary
-        participation.setInitializationState(INITIALIZED);
+
+        // If the instructor retests a participation after the deadline set the state back to finished. Otherwise, the participation is initialized
```
Comment (julian-christl) [inline src/main/java/de/tum/in/www1/artemis/service/ParticipationService.java:354]: Explain why this doesn't affect test runs in the comment.

**Review by julian-christl — COMMENTED** [review]
Tested on TS1, works now.

CR one small comment

**Review by ge65cer — DISMISSED** [review]
My previous concerns were addressed. Nice job!

File: src/main/java/de/tum/in/www1/artemis/service/ParticipationService.java (hunk 2 above)
Comment (JohannesStoehr) [reply src/main/java/de/tum/in/www1/artemis/service/ParticipationService.java:354]: An exercise in an exam does not have a deadline on its own. So the comment already describes the behavior correctly.

**Review by julian-christl — APPROVED** [review]
Reapprove. The fix only extends the scope, so the functionality itself didn't change.

**Review by ge65cer — APPROVED** [review]
Reapprove

**Review by stefanwaldhauser — APPROVED** [review]
LGTM Approved :)

**Pull request merged**